### PR TITLE
Organization Mapping Zip Codes

### PR DESCRIPTION
**Pull request by emichaud998**
Changes in this PR:

Changes have been made to allow zip code information found in some endpoint lists to be used in endpoint to organization mapping. Major changes include:
1. Updated endpoint list parsers to extract organization zip code information if available and save it in db
2.  Splitting organization information from fhir_endpoints table into its own table and making a map table that linked the fhir_endpoints table to this fhir_endpoint_organizations table
3. Updated the adding, updating, and removal of fhir endpoint information in the fhir_endpoint table to work with the new fhir_endpoint_organizations table
4. Updated the endpoint linker algorithm to use zip codes when making matches 

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
- [x] Administrative tasks are complete.
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: [Lantern 545 JIRA Ticket](https://oncprojectracking.healthit.gov/support/browse/LANTERN-545)
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
- [x] Code review has been performed.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

### CODE REVIEW

File: endpointmanager/pkg/endpointmanager/fhirendpoint_test.go [hunk 1]
```diff
@@ -141,17 +156,17 @@ func Test_FHIREndpointEqual(t *testing.T) {
...
-	endpoint2.OrganizationNames = endpoint1.OrganizationNames
+	endpoint2.OrganizationList[0].OrganizationName = endpoint1.OrganizationList[0].OrganizationName
 
-	endpoint2.NPIIDs = []string{"Other 1"}
+	endpoint2.OrganizationList[0].OrganizationNPIID = "Other 1"
 	if endpoint1.Equal(endpoint2) {
 		t.Error("Did not expect endpoint1 to equal endpoint 2. NPIIDs should be different.")
 	}
-	endpoint2.NPIIDs = endpoint1.NPIIDs
+	endpoint2.OrganizationList[0].OrganizationNPIID = endpoint1.OrganizationList[0].OrganizationNPIID
```
Comment (bmath10) [inline endpointmanager/pkg/endpointmanager/fhirendpoint_test.go:170]: I think it would be good to add some other edge case tests of the `OrganizationListEquals` function so that we verify that if the organizations are the same but in a different order, it doesn't register that as a change, and if the lists have the same organizations, but one has a few extra organizations, that is also a change.

File: endpointmanager/pkg/endpointmanager/fhirendpoint_test.go [hunk 2]
```diff
@@ -175,63 +190,3 @@ func Test_FHIREndpointEqual(t *testing.T) {
 		t.Errorf("Nil endpoint 1 should equal nil endpoint 2.")
 	}
 }
-
-func Test_AddOrganizationName(t *testing.T) {
```
Comment (bmath10) [inline endpointmanager/pkg/endpointmanager/fhirendpoint_test.go:179]: Why were these tests removed?

**Review by bmath10 — CHANGES_REQUESTED** [review]
Just started reviewing so probably won't be all of the comments I have!

File: endpointmanager/pkg/endpointmanager/fhirendpoint_test.go (hunk 2 above)
Comment (emichaud998) [reply endpointmanager/pkg/endpointmanager/fhirendpoint_test.go:179]: This function is no longer used since we don't need to add organizations to the existing organization list so that the entire array with the new entries can be added to the database. Now, new organizations are added by checking to see if an entry exists in the db, and if not, it adds the new entry, otherwise it just updates the update time. Therefore, since this function is no longer used and was removed, the tests were removed!

File: endpointmanager/pkg/endpointmanager/fhirendpoint_test.go (hunk 1 above)
Comment (emichaud998) [reply endpointmanager/pkg/endpointmanager/fhirendpoint_test.go:170]: Added two new tests, one that checks to make sure endpoint1 still equals endpoint2 even if their organization lists are not in the same order, and one that checks to make sure endpoint1 does not equal endpoint2 if one of them has more organizations in their organization list than the other

File: db/sql/dbsetup.sql [hunk 3]
```diff
@@ -359,6 +380,10 @@ RIGHT JOIN fhir_endpoints AS endpts ON links.url = endpts.url
 LEFT JOIN fhir_endpoints_info AS endpts_info ON endpts.url = endpts_info.url   
 LEFT JOIN vendors ON endpts_info.vendor_id = vendors.id
 LEFT JOIN npi_organizations AS orgs ON links.organization_npi_id = orgs.npi_id
+LEFT JOIN (SELECT fom.id as id, array_agg(fo.organization_name) as organization_names
```
Comment (bmath10) [inline db/sql/dbsetup.sql:383]: Since these 4 lines (383-386) are repeated in this file twice and at least one other place (archivefile), I'm wondering if it could be made into a separate view, and then that view could be used elsewhere? i'm not sure if you can call a view in a view, but I feel like it would improve efficiency if this complicated query only had to be done once in a separate view.

Comment (bmath10) [reply db/sql/dbsetup.sql:383]: Also on the topic of efficiency, it might be worth adding indices for the new reference ids? I'm not entirely sure how indices work, but it looks like we have them for most ids that are used across tables. We could also add them for the validations table too since i'm not seeing those either

File: endpointmanager/pkg/chplendpointquerier/NextGenwebscraper.go [hunk 4]
```diff
@@ -21,21 +21,19 @@ func NextGenwebscraper(CHPLURL string, fileToWriteTo string) {
...
-					entryDSTU2.OrganizationName = organizationName
-					entryDSTU2.URL = DSTU2URL
+					zipCode := strings.TrimSpace(tableEntries.Eq(5).Text())
+					URL := strings.TrimSpace(tableEntries.Eq(6).Text())
 
-					entryR4.OrganizationName = organizationName
-					entryR4.URL = R4URL
+					entry.OrganizationName = organizationName
+					entry.URL = URL
+					entry.OrganizationZipCode = zipCode
```
Comment (bmath10) [inline endpointmanager/pkg/chplendpointquerier/NextGenwebscraper.go:34]: It looks like **some** of the NextGen zip code is 9 characters, do we cut the zipcodes down to 5 characters here or somewhere else?

**Review by bmath10 — CHANGES_REQUESTED** [review]

File: endpointmanager/pkg/endpointmanager/fhirendpoint.go [hunk 5]
```diff
@@ -78,17 +128,54 @@ func NormalizeEndpointURL(url string) string {
...
+	var OrganizationNames []string
+	for _, org := range e.OrganizationList {
+		OrganizationNames = append(OrganizationNames, org.OrganizationName)
+	}
+
+	return OrganizationNames
+}
+
+// AddOrganization adds the organization to the endpoint's Organization list if it's not present already. If it is, it does nothing.
+func (e *FHIREndpoint) AddOrganization(org *FHIREndpointOrganization) {
```
[the remark this answers is not in the document]
Comment (emichaud998) [reply endpointmanager/pkg/endpointmanager/fhirendpoint.go:169]: Nope not anymore actually, I will remove it!

File: endpointmanager/pkg/chplendpointquerier/NextGenwebscraper.go (hunk 4 above)
Comment (emichaud998) [reply endpointmanager/pkg/chplendpointquerier/NextGenwebscraper.go:34]: I didn't notice this! I just added a check to the Lantern format fetcher parser that will cut down any zip codes longer than 5 characters down to 5!

**Comment by bmath10** [comment]
README in db folder should be updated with database changes

**Review by bmath10 — APPROVED** [review]
